### PR TITLE
Include project dependencies in Checkstyle classpath

### LINKED ISSUE #855: Checkstyle classpath is set incorrectly for multi-project builds
### Expected Behavior
If project A depends on project B, then the javadoc for a function defined in A should be able to reference an exception defined in B.

### Current Behavior
* If a function's javadoc has more than one `@throws` tag, one of which references an exception in a different project, then the Gradle Checkstyle task shows an error:
```
[ant:checkstyle] [ERROR] .../checkstyle-bug/client/src/main/java/csbug/BarInterface.java:10:22: Unable to get class information for @throws tag 'FooException'. [JavadocMethod]
```
* This doesn't happen if there's only a single `@throws` tag
* This only happens for Checkstyle newer than 6.11.2 (including the most recent release 7.2)
* This happens for Gradle 2.12 - 3.1
* This can be fixed by manually adding dependencies to the Checkstyle classpath:
```
tasks.checkstyleMain.classpath += configurations.compileClasspath
```

### Context
It's reasonable to have a project that contains API classes (interfaces, data formats, exception types) and dependent projects that reference API classes. Functions that throw (and document) multiple exception types are common as well.

### Steps to Reproduce (for bugs)
https://github.com/punya/checkstyle-bug minimally demonstrates the problem. [Uncomment these two lines](https://github.com/punya/checkstyle-bug/blob/bf22270d/build.gradle#L23-L24) to fix the problem manually.

### Your Environment
 * Build scan URL: https://scans.gradle.com/s/umlviaz4noymq

### PR DESCRIPTION
**Pull request by punya**
Fixes #855.

Any of the checked boxes below indicate that I took action:

- [x] Reviewed the [Contribution Guidelines](https://github.com/gradle/gradle/blob/master/.github/CONTRIBUTING.md).
- [x] Signed the [Gradle CLA](http://gradle.org/contributor-license-agreement/).
- [x] Ensured that basic checks pass: `./gradlew quickCheck`

For all non-trivial changes that modify the behavior or public API:

- [x] Before submitting a pull request, I started a discussion on the [Gradle developer list](https://groups.google.com/forum/#!forum/gradle-dev),
      the [forum](https://discuss.gradle.org/) or can reference a [JIRA issue](https://issues.gradle.org/secure/Dashboard.jspa).
- [x] I considered writing a design document. A design document can be
brief but explains the use case or problem you are trying to solve,
touches on the planned implementation approach as well as the test cases
that verify the behavior. Optimally, design documents should be submitted
as a separate pull request. [Samples](https://github.com/gradle/gradle/tree/master/design-docs)
can be found in the Gradle GitHub repository. Please let us know if you need help with
creating the design document. We are happy to help!
- [x] The pull request contains an appropriate level of unit and integration
test coverage to verify the behavior. Before submitting the pull request
I ran a build on your local machine via the command
`./gradlew quickCheck <impacted-subproject>:check`.
- [x] The pull request updates the Gradle documentation like user guide,
DSL reference and Javadocs where applicable.

### CODE REVIEW

**Comment by punya** [comment]
@eriwen this is ready for review.

**Comment by punya** [comment]
@eriwen sorry to be a bother, but when do you think this can get reviewed?

**Comment by eriwen** [comment]
@punya I make no hard promises, but I'm hoping to address some pull requests tomorrow. Is this an urgent need?

**Comment by punya** [comment]
Thanks @eriwen, that would be great! It's not urgent, I just wanted to make
sure this didn't get buried/lost.

**Comment by eriwen** [comment]
@punya I'm afraid I don't see your CLA :( — Is it under a different name than Punyashloka Biswal? If not, can you please submit another one?

**Comment by punya** [comment]
I'm surprised - I thought I signed the CLA for an earlier PR. I'll work with my employer and get this done today.

**Comment by punya** [comment]
@eriwen I've signed the CLA now, sorry about the confusion earlier.

File: subprojects/code-quality/src/integTest/groovy/org/gradle/api/plugins/quality/CheckstylePluginClasspathIntegrationTest.groovy [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+@Issue("gradle/gradle#855")
+class CheckstylePluginClasspathIntegrationTest extends AbstractIntegrationSpec {
+    def setup() {
+        writeBuildFile()
+        writeConfigFile()
+        goodCode()
+    }
+
+    def "accepts throwing exception from other project"() {
+        succeeds("checkstyleMain")
```
Comment (eriwen) [inline subprojects/code-quality/src/integTest/groovy/org/gradle/api/plugins/quality/CheckstylePluginClasspathIntegrationTest.groovy]: You need `expect:` on the line above to make this test run using Spock

File: subprojects/code-quality/src/integTest/groovy/org/gradle/api/plugins/quality/CheckstylePluginClasspathIntegrationTest.groovy [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        file("build.gradle") << """
+subprojects {
+    apply plugin: "java"
+    apply plugin: "checkstyle"
+
+    repositories {
+        mavenCentral()
+    }
+
+    checkstyle.toolVersion = "0.7.2"
```
Comment (eriwen) [inline subprojects/code-quality/src/integTest/groovy/org/gradle/api/plugins/quality/CheckstylePluginClasspathIntegrationTest.groovy]: Tool version should be 7.2

Comment (eriwen) [inline subprojects/code-quality/src/integTest/groovy/org/gradle/api/plugins/quality/CheckstylePluginClasspathIntegrationTest.groovy]: Need to also configure a common checkstyle.xml location here like in your example project

File: subprojects/code-quality/src/integTest/groovy/org/gradle/api/plugins/quality/CheckstylePluginClasspathIntegrationTest.groovy [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+class Exn extends Exception { }
+        """
+
+        file("client/src/main/java/csbug/Iface.java") << """
+package csbug;
+
+interface Iface {
+    /**
+     * @throws Exn
```
Comment (eriwen) [inline subprojects/code-quality/src/integTest/groovy/org/gradle/api/plugins/quality/CheckstylePluginClasspathIntegrationTest.groovy]: Checkstyle doesn't pick this declaration up for some reason.

**Comment by eriwen** [comment]
This is merged and polished with https://github.com/gradle/gradle/commit/966e74fcf9ea51d19c3df99703b0a96fa868b8bb

Thanks @punya!

**Comment by punya** [comment]
Thanks for fixing my mistakes - I'll try to do better next time!

**Comment by eriwen** [comment]
Hey now, you did everything exactly right except the test. That is far
better than most - thank you!

**Pull request closed without merging**